src/topology/algebra/filter_basis.lean

```diff
@@ -0,0 +1,400 @@
+/-
+Copyright (c) 2021 Patrick Massot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Patrick Massot
+-/
+
+import order.filter.bases
+import topology.algebra.module
+/-!
+# Group and ring filter bases
+
+A `group_filter_basis` is a `filter_basis` on a group with some properties relating
+the basis to the group structure. The main theorem is that a `group_filter_basis`
+on a group gives a topology on the group which makes it into a topological group
+with neighborhoods of the neutral element generated by the given basis.
+
+## Main definitions and results
+
+Given a group `G` and a ring `R`:
+
+* `group_filter_basis G`: the type of filter bases that will become neighborhood of `1`
+  for a topology on `G` compatible with the group structure
+* `group_filter_basis.topology`: the associated topology
+* `group_filter_basis.is_topological_group`: the compatibility between the above topology
+  and the group structure
+* `ring_filter_basis R`: the type of filter bases that will become neighborhood of `0`
+  for a topology on `R` compatible with the ring structure
+* `ring_filter_basis.topology`: the associated topology
+* `ring_filter_basis.is_topological_ring`: the compatibility between the above topology
+  and the ring structure
+
+## References
+
+* [N. Bourbaki, *General Topology*][bourbaki1966]
+-/
+
+open filter set topological_space function
+open_locale topological_space filter pointwise
+
+universe u
+
+/-- A `group_filter_basis` on a group is a `filter_basis` satisfying some additional axioms.
+  Example : if `G` is a topological group then the neighbourhoods of the identity are a
+  `group_filter_basis`. Conversely given a `group_filter_basis` one can define a topology
+  compatible with the group structure on `G`.  -/
+class group_filter_basis (G : Type u) [group G] extends filter_basis G :=
+(one' : ∀ {U}, U ∈ sets → (1 : G) ∈ U)
+(mul' : ∀ {U}, U ∈ sets → ∃ V ∈ sets, V * V ⊆ U)
+(inv' : ∀ {U}, U ∈ sets → ∃ V ∈ sets, V ⊆ (λ x, x⁻¹) ⁻¹' U)
+(conj' : ∀ x₀, ∀ {U}, U ∈ sets → ∃ V ∈ sets, V ⊆ (λ x, x₀*x*x₀⁻¹) ⁻¹' U)
+
+/-- A `add_group_filter_basis` on an additive group is a `filter_basis` satisfying some additional
+  axioms. Example : if `G` is a topological group then the neighbourhoods of the identity are a
+  `add_group_filter_basis`. Conversely given a `add_group_filter_basis` one can define a topology
+  compatible with the group structure on `G`. -/
+class add_group_filter_basis (A : Type u) [add_group A] extends filter_basis A :=
+(zero' : ∀ {U}, U ∈ sets → (0 : A) ∈ U)
+(add' : ∀ {U}, U ∈ sets → ∃ V ∈ sets, V + V ⊆ U)
+(neg' : ∀ {U}, U ∈ sets → ∃ V ∈ sets, V ⊆ (λ x, -x) ⁻¹' U)
+(conj' : ∀ x₀, ∀ {U}, U ∈ sets → ∃ V ∈ sets, V ⊆ (λ x, x₀+x+-x₀) ⁻¹' U)
```

Comment:
Is this better or worse as
```suggestion
(conj' : ∀ x₀, ∀ {U}, U ∈ sets → ∃ V ∈ sets, V ⊆ mul_aut.conj x₀ ⁻¹' U)
```

Reply:
Perhaps as `add_aut`? :-) I don't think it matters much.

Reply:
I don't see the point of obfuscating this.